### DISCUSSION TITLE
DietPi-Software | Zerotier

### DISCUSSION THREAD

**Discussion by cocoflan** [MichaIng/DietPi#5335] [Software Request] [answered]
#### Formal software information
- Software title | Zerotier
- Short description | People use ZeroTier to seamlessly connect laptops, desktops, phones, embedded devices, cloud resources, and apps any way they want anywhere they go. Our software transforms the entire world into a single data center.
- Official URL (if available) | https://zerotier.com/
- GitHub URL (if available) | https://github.com/zerotier/ZeroTierOne
- (Official) install documentation | https://zerotier.com/download.shtml

#### Can you provide the installation steps that you would suggest DietPi-Software to do?

curl -s 'https://raw.githubusercontent.com/zerotier/download.zerotier.com/master/htdocs/contact%40zerotier.com.gpg' | gpg --import && \
if z=$(curl -s 'https://install.zerotier.com/' | gpg); then echo "$z" | sudo bash; fi

Some users may need to replace gpg with gpg2 in the above commands.

This first makes sure you have the contact@zerotier.com GPG key. If this succeeds it then fetches our install script and pipes it through GPG. This verifies the script against its embedded signature, and if it is valid only the verified portion is forwarded through to "sudo bash." Otherwise the script is not executed and you will see a GPG error about signature check failure.

The install.zerotier.com script first characterizes your Linux distribution and adds the ZeroTier official package repository if we have one available for you. It then installs the zerotier-one package, enables and starts the service, and prints your new 10-digit ZeroTier address.

For those who are a bit more careless lazy brave, the script also works without GPG:

curl -s https://install.zerotier.com/ | sudo bash

#### Are you willing to help maintaining the software installation, e.g. in case of needed setup changes due to updates etc.? This is not needed, but could speed up our decision to implement it, as man power is always a topic :wink:.

Yes, I could help out with this, e.g. if I experience issues. / Sorry, I am totally unexperienced in bash coding. -->

**Comment by MichaIng** [comment]
@cocoflan 
Many thanks for your request. Definitely an interesting project if you want to connect two peers without having everything relayed over a VPN.

**Comment by cocoflan** [comment]
I am testing it on my Raspberry pi 3 and Nanopi M4 for weeks now, all works perfect!

Greetings, hope it will be in dietpi soon.

**Comment by JimmyRecard** [comment]
I would love this as well!

**Comment by cocoflan** [comment]
Any news if Zerotier could be added to the  packages ?

**Comment by MichaIng** [comment]
@cocoflan 
Probably with DietPi 6.30. However if someone is able to do a start, this could be done generally faster.

**Comment by MichaIng** [comment]
@KalixtShawxo
> I have been trying to add ZeroTier VPN to a DietPi system and repeatedly it fails when it comes to cleaning up the OS for final launch. I had it running on a RasPi 3 A+ fin with regular raspian lite 32-bit install for the longest time. then I tried it on DietPi a few times, keep running into the same problem over and over. I asked for support at ZeroTier and it looks like an OS issue. I have now resetup the ZeroTier successfully using the Raspian lite 64-bit on the same RasPi 3 A+ and it is working fine. I would really like to see it on DietPi, since most all my other SBCs run DietPi except for the two running OpenWRT.
> 
> I am not smart enough to make it work with DietPi, but the software works fantastically on raspian for remoting into my network and allowing me to use my cellphone, tablet, and laptop away from home as if I am at home.... including SSH to my various SBCs.
> 
> Disclaimer: I have no affiliation with ZeroTier other than being a satisfied user, and thought it would be a good add-on to the DietPi-Software.
> 
> Thanks.

Moved your post here to avoid doubled discussion. Error messages are always helpful to help diagnose the issue 😉.

**Comment by MichaIng** [comment]
For completeness, this had 52 votes over 3 years on FeatHub: https://feathub.com/MichaIng/DietPi/+70
Please place your votes here on OP again, since FeatHub is dead.

**Comment by KalixtShawxo** [comment]
from the Zerotier community post i made:

 "long story short… when I get to he installation part where you remove all the “existing network stuff”, it locks up and stops at 19%. I redid the installation several times, and each time it stopped at 19% of the removal after issuing this command:
sudo apt remove --purge --auto-remove dhcpcd5 fake-hwclock ifupdown isc-dhcp-client isc-dhcp-common openresolv

So after the several tries with each one being done to a clean 16GB Sandisk Ultra MicroSDHC card, where each one locked up at the same point, and reboots of the Raspi would not load anything, I redid a clean install and omitted the removal step and the installation was able to complete, with all the necessary configuration completed to where the bridge does now work.
So here is the issue… with those packages (dhcpcd5 fake-hwclock ifupdown isc-dhcp-client isc-dhcp-common openresolv) still installed and not removed… "
here's the update....with the packages left installed, it appears to run on the Pi, but it will not allow communication from a remote system connected to the VPN. I had thought it was working  testing it locally.....but when at my daughter's house and trying to get my son-in-law connected nothing would communicate across the network.the control page on the ZeroTier website showed the bridge in place online, and my phone, tablet, laptop, and my SIL's laptop all connected to the VPN but nothing was bridged. i kept trouble shooting anything else to see what it could be, and eventually gave up and did the new 64bit Raspian.     On Saturday, March 5, 2022, 09:21:01 AM EST, MichaIng ***@***.***> wrote:  
 
 


For completeness, this had 52 votes over 3 years on FeatHub: https://feathub.com/MichaIng/DietPi/+70

**Comment by Joulinar** [comment]
Basically removing `ifupdown` could be an issue as this is an essential component of DietPi to manage network. Removing this  will most probably break your system.

**Comment by KalixtShawxo** [comment]
and that's why you're the programmer genius..... you spotted that right off the bat.  me....not so much.  :)
so that begs the question, is it something that would work if only that was left in place,,,, or woulld something else have to be done to make it work?and then the big question is it something that could be integrated with the DietPi-Software catalog so that all could have the option to use it? 
Thanks for your help so far.      On Saturday, March 5, 2022, 10:10:54 AM EST, Joulinar ***@***.***> wrote:  
 
 


Basically removing ifupdown could be an issue as this is an essential component of DietPi to manage network. Removing this will most probably break your system.

**Comment by Joulinar** [comment]
Not sure how ZeroTier is working. But it is a showstopper if their install scripts us trying to remove `ifupdown`.

**Comment by KalixtShawxo** [comment]
here is the link to the bridge creation instructions;
https://zerotier.atlassian.net/wiki/spaces/SD/pages/193134593/Bridge+your+ZeroTier+and+local+network+with+a+RaspberryPi

the remove ifupdown is not automated, it is part of a manual command. 
<sudo apt remove --purge --auto-remove dhcpcd5 fake-hwclock ifupdown isc-dhcp-client isc-dhcp-common openresolv> 
knowing now that that may be part of the problem (or all of the problem potentially) then maybe an installation with that package left in place by instead running is as ;
<sudo apt remove --purge --auto-remove dhcpcd5 fake-hwclock isc-dhcp-client isc-dhcp-common openresolv> 

maybe then it would work.... I know from past experiments that leaving them all in place will not work. (not running the remove command at all) then there could be something else that needs to be done... I dunno, that's why i asked about it. maybe remove it and reinstall it?  just spitballing here. 

thoughts?    On Saturday, March 5, 2022, 10:54:34 AM EST, Joulinar ***@***.***> wrote:  
 
 


Not sure how ZeroTier is working. But it is a showstopper if their install scripts us trying to remove ifupdown.

**Comment by MichaIng** [comment]
That whole think to switch to `systemd-networkd` shouldn't be required. It would be a deal breaker for Zerotier indeed when it forced users to use `systemd-networkd` or systemd at all 😄. Please check out guides about how to setup a network bridge via `/etc/network/interfaces` (but use a dedicated file like `/etc/network/interfaces.d/br0.conf`).

**Comment by cocoflan** [comment]
Maybe https://tailscale.com/, is better for the job, i have tested it for month's now. also on the feathub page, https://feathub.com/MichaIng/DietPi/+226. 
I know i was the one that wanted zerotier to be implemented via feathub, but tailscale is much better.

32bit and 64bit compatible. Works perfect on all my systems 32bit and 64 bit.


https://tailscale.com/kb/1139/tailscale-vs-zerotier/

https://github.com/MichaIng/DietPi/issues/4893

**Comment by KalixtShawxo** [comment]
Guess I will need to take a look at tailscale then. Thanks for the information. 👍👌😃

Sent from Yahoo Mail on Android 
 
  On Sat, Mar 5, 2022 at 20:02, ***@***.***> wrote:   


Maybe https://tailscale.com/, is better for the job, i have tested it for month's now. also on the feathub page, https://feathub.com/MichaIng/DietPi/+226.
I know i was the one that wanted zerotier to be implemented via feathub, but tailscale is much better.

32bit and 64bit compatible. Works perfect on all my systems 32bit and 64 bit.

**Comment by Joulinar** [comment]
Using Wireguard is the easiest thing I could imagine to connect all my device back to my home network.

**Comment by KalixtShawxo** [comment]
The reason i went with zerotier was the bridge being so easy to initially setup, before i tried it with DietPi.... since that allowed me to connect to literally everything on my home network without installing somethjng on every device i wished to access. Things like NVRs, cameras, and other devices that i can't install client software, were still accessable now without having to have any port forwarding through my router like i used to have to use. Now my router and firewall are completely stealthed to the outside world, but i can access it all remotely, not just PCs or SBCs that i can install software on. 
Doesn't look like the others based on wireguard has that, not that i found yet in my initial reviews so far anyway. 
But, great information here... shame that ZeroTier won't be DietPi compatible, thanks to all for your input. It is always appreciated as i learn morr. 

Sent from Yahoo Mail on Android 
 
  On Sun, Mar 6, 2022 at 1:55, ***@***.***> wrote:   


Using Wireguard is the easiest thing I could imagine to connect all my device back to my home network.

**Reply by Joulinar** [reply]
Using a VPN will allow to access every device inside your local network. Doesn't matter what they are. You just need the VPN server installed on a single device and the client app that can be installed on every mobile phone from the available software store. In meantime, even some router offer VPN server functionality like Wireguard directly.

**Comment by KalixtShawxo** [comment]
I know what you are saying, part of the problem is I was looking at VPN options but they seem to all require opening ports to your home network, or getting certificates, or lots of settings/configuration, etc.... whereas ZeroTier did not... It connected out from the bridge (RasPi 3 A+)  though my firewall/router to ZT server which I then connected to from whatever remote device I used, and then that external connection allowed me to connect to the existing bridge and thus my network.  No certificates, no open ports, just build the bridge and connect. it really was so easy an idiot like me could do it. The issues only came up when I tried to rebuild it on a DietPi based system. I'm learning slowly, lots of experimenting with the easy stuff.  DietPi is just friendlier overall so I had thought I'd combine the two... Not knowing (what I know now) about the conflicting requirements. 

Still looking for simple ones compatible with DietPi, or I may just leave it on ZeroTier with the Raspbian since it works well, and just do without having the added DietPi niceties. 
Thanks again for the responses, and the information. 
Sent from Yahoo Mail on Android 
 
  On Sun, Mar 6, 2022 at 7:53, ***@***.***> wrote:   


Using a VPN will allow to access every device inside your local network. Doesn't matter what they are. You just need the VPN server installed on a single device and the client app that can be installed on every mobile phone from the available software store. In meantime, even some router offer VPN server functionality like Wireguard directly.

**Reply by Joulinar** [reply]
Using Wireguard on DietPi will bring up a fully configured VPN server ootb. Nothing like certificated needs to be created, installed or setup. Just a single UDP port to be open/forward. Use `dietpi-software` to install the server and scan the QR code with your mobile. That's it. Ok using  a DDNS would be helpful to use as well.

**Comment by KalixtShawxo** [comment]
Looking at options with "tailscale" now as it was mentioned above... Looks like it is an enhanced wireguard closer to ZeroTier in it's setup. 

Sent from Yahoo Mail on Android 
 
  On Sun, Mar 6, 2022 at 9:36, Roger ***@***.***> wrote:   I know what you are saying, part of the problem is I was looking at VPN options but they seem to all require opening ports to your home network, or getting certificates, or lots of settings/configuration, etc.... whereas ZeroTier did not... It connected out from the bridge (RasPi 3 A+)  though my firewall/router to ZT server which I then connected to from whatever remote device I used, and then that external connection allowed me to connect to the existing bridge and thus my network.  No certificates, no open ports, just build the bridge and connect. it really was so easy an idiot like me could do it. The issues only came up when I tried to rebuild it on a DietPi based system. I'm learning slowly, lots of experimenting with the easy stuff.  DietPi is just friendlier overall so I had thought I'd combine the two... Not knowing (what I know now) about the conflicting requirements. 

Still looking for simple ones compatible with DietPi, or I may just leave it on ZeroTier with the Raspbian since it works well, and just do without having the added DietPi niceties. 
Thanks again for the responses, and the information. 
Sent from Yahoo Mail on Android 
 
  On Sun, Mar 6, 2022 at 7:53, ***@***.***> wrote:   


Using a VPN will allow to access every device inside your local network. Doesn't matter what they are. You just need the VPN server installed on a single device and the client app that can be installed on every mobile phone from the available software store. In meantime, even some router offer VPN server functionality like Wireguard directly.

**Comment by KalixtShawxo** [comment]
Michalng, thanks so much for that info. Funny thing is I just got done setting up TailScale as a replacement for ZeroTier. I pulled the SD card from my Rpi 3A+, put in a new one with a fresh load of DietPi, ran the install, set the subnets, set the forwarding, etc,  and it works. No router or firewall changes needed. 
You could maybe consider adding TailScale to the Dietpi-software catalog alongside the Wireguard option, since ZeroTier is a No-Go. 
Again, thanks for your help and information, I just did a PayPal donation to DietPi as a thank you for great software and your support. 


Sent from Yahoo Mail on Android 
 
  On Sun, Mar 6, 2022 at 17:00, ***@***.***> wrote:   


Using Wireguard on DietPi will bring up a fully configured VPN server ootb. Nothing like certificated needs to be created, installed or setup. Just a single UDP port needs to be open. Use dietpi-software to install the server and scan the QR code with your mobile. That's it. Ok using a DDNS would be helpful to use as well.

**Reply by MichaIng** [reply]
An argument pro Zerotier is that it is better known and wider used, so that e.g. security issues are recognised faster, bugs are fixed sooner etc. It is not a no-go, since the network bridge can be created with ifupdown and /etc/network/interfaces as well. Both projects have their value, alternatives are good to have.

**Comment by KalixtShawxo** [comment]
I totally misunderstood the ifupdown removal for ZeroTier as being a No-Go cause then,  I don't know what workarounds are needed to make it work with DietPi, that's why I suggested bringing it into the Dietpi-software catalog so that it could be kitted out to be compatible. I would prefer ZeroTier over TailScale for the fact that it's easier to share things with my son-in-law on the network as a whole instead of individual devices. Nonetheless, I switched over and got TailScale working under DietPi quite easily and I am the sole user 99.5% of the time, so really as long as I can access my home network from wherever, and using DietPi (for consistency) then I am happy. 
Thanks again so much for your time and knowledge. Roger.

Sent from Yahoo Mail on Android 
 
  On Mon, Mar 7, 2022 at 11:12, ***@***.***> wrote:   


An argument pro Zerotier is that it is better known and wider used, so that e.g. security issues are recognised faster, bugs are fixed sooner etc. It is not a no-go, since the network bridge can be created with ifupdown and /etc/network/interfaces as well. Both projects have their value, alternatives are good to have.

**Comment by MichaIng** [comment]
> I totally misunderstood the ifupdown removal for ZeroTier as being a No-Go cause then

`ifupdown` does not need to be removed since the network bridge can be created with it as well, instead of using `systemd-networkd` 🙂.

**Comment by Caporeira** [comment]
Someone had already installed ZeroTier on DietPi ?
I have no public IP, that way WireGuard is not for me.
I'm using ZT on few devices and i with love to have ZT on DietPi - but i don't want to damage DietPi.

**Reply by Joulinar** [reply]
you should be able to do following to install ZeroTier. Btw, this will not remove anything at this point. 

```
eval "curl -sSLf https://raw.githubusercontent.com/zerotier/ZeroTierOne/master/doc/contact%40zerotier.com.gpg | gpg --dearmor -o /etc/apt/trusted.gpg.d/dietpi-zerotier.gpg --yes"
eval "echo 'deb https://download.zerotier.com/debian/bullseye bullseye main' > /etc/apt/sources.list.d/zerotier.list"
apt update
apt install zerotier-one
```

Honestly I don't use ZeroTier and I'm not sure if it is working out of the box afterwards. I guess you could join some network. But if I understood correctly, ZeroTier docs are stating to remove some network packages to allow to bridge your ZeroTier and local network with a RaspberryPi. Means someone would need to check if the a `br0` interface could be used together with `ifupdown` as well or if there is a strict need to use `systemd-networkd`.

https://zerotier.atlassian.net/wiki/spaces/SD/pages/193134593/Bridge+your+ZeroTier+and+local+network+with+a+RaspberryPi
https://wiki.debian.org/BridgeNetworkConnections

**Comment by AAS-Crypt** [comment]
Have been using ZT for months now, great software. Works right out of box, would recommend to use it for everyone. Yet I'm little dissapointed that they dissmissed the whole "Moons" thing. Would have been great for moments of Countrywide, Citywide or even ISPwide external internet cuting off.

**Comment by Joulinar** [comment, marked as answer]
will be part of DietPi v8.6
